File: core/src/nl/t64/game/rpg/audio/AudioManager.java
package nl.t64.game.rpg.audio;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.audio.Music;
import com.badlogic.gdx.audio.Sound;
import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.constants.Constant;

import java.util.HashMap;
import java.util.List;
import java.util.Map;


public class AudioManager {

    private static final float BGM_VOLUME = 0.2f;
    private static final float BGS_VOLUME = 0.5f;

    private final Map<String, Music> queuedBgm;
    private final Map<String, Music> queuedBgs;
    private final Map<String, Sound> queuedMe;
    private final Map<String, Sound> queuedSe;

    public AudioManager() {
        this.queuedBgm = new HashMap<>();
        this.queuedBgs = new HashMap<>();
        this.queuedMe = new HashMap<>();
        this.queuedSe = new HashMap<>();
    }

    public void toggleMusic(boolean isMusicOn, boolean mustPlayBgmImmediately) {
        if (isMusicOn && mustPlayBgmImmediately) {
            handle(AudioCommand.BGM_PLAY_LOOP, AudioEvent.BGM_TITLE);
        } else {
            handle(AudioCommand.BGM_STOP_ALL);
        }
    }

    public void toggleSound(boolean isSoundOn) {
        if (isSoundOn) {
            handle(AudioCommand.SE_PLAY_ONCE, AudioEvent.SE_MENU_CONFIRM);
        } else {
            handle(AudioCommand.SE_STOP, AudioEvent.SE_MENU_CONFIRM);
        }
    }

    public void possibleBgmFade(AudioEvent currentBgm, AudioEvent newBgm) {
        if (currentBgm != newBgm) {
            queuedBgm.values().forEach(bgm -> fade(bgm, BGM_VOLUME));
        }
    }

    public void possibleBgsFade(List<AudioEvent> currentBgs, List<AudioEvent> newBgs) {
        currentBgs.stream()
                  .filter(event -> !event.equals(AudioEvent.NONE))
                  .filter(event -> !newBgs.contains(event))
                  .forEach(event -> fade(queuedBgs.get(event.filePath), BGS_VOLUME));
    }

    public void possibleBgmSwitch(AudioEvent prevBgm, AudioEvent nextBgm) {
        if (prevBgm != nextBgm) {
            handle(AudioCommand.BGM_STOP_ALL);
            handle(AudioCommand.BGM_PLAY_LOOP, nextBgm);
        }
    }

    public void possibleBgsSwitch(List<AudioEvent> prevBgs, List<AudioEvent> nextBgs) {
        prevBgs.stream()
               .filter(event -> !event.equals(AudioEvent.NONE))
               .filter(event -> !nextBgs.contains(event))
               .forEach(event -> handle(AudioCommand.BGS_STOP, event));
        nextBgs.stream()
               .filter(event -> !event.equals(AudioEvent.NONE))
               .filter(event -> !prevBgs.contains(event))
               .forEach(event -> handle(AudioCommand.BGS_PLAY_LOOP, event));
    }

    public void fadeBgmBgs() {
        queuedBgm.values().forEach(bgm -> fade(bgm, BGM_VOLUME));
        queuedBgs.values().forEach(bgs -> fade(bgs, BGS_VOLUME));
    }

    public void handle(AudioCommand command, List<AudioEvent> events) {
        events.forEach(event -> handle(command, event));
    }

    public void handle(AudioCommand command, AudioEvent event) {
        switch (command) {
            case BGM_PLAY_ONCE -> playBgm(event.filePath, false);
            case BGM_PLAY_LOOP -> playBgm(event.filePath, true);
            case BGM_STOP -> queuedBgm.get(event.filePath).stop();
            case BGM_PAUSE -> queuedBgm.get(event.filePath).pause();

            case BGS_PLAY_ONCE -> playBgs(event.filePath, false);
            case BGS_PLAY_LOOP -> playBgs(event.filePath, true);
            case BGS_STOP -> queuedBgs.get(event.filePath).stop();
            case BGS_PAUSE -> queuedBgs.get(event.filePath).pause();

            case ME_PLAY_ONCE -> playMe(event.filePath, false);
            case ME_PLAY_LOOP -> playMe(event.filePath, true);
            case ME_STOP -> queuedMe.get(event.filePath).stop();

            case SE_PLAY_ONCE -> playSe(event, false);
            case SE_PLAY_LOOP -> playSe(event, true);
            case SE_STOP -> queuedSe.get(event.filePath).stop();
            default -> throw new IllegalArgumentException("Call 'ALL' AudioCommands without second argument.");
        }
    }

    public void handle(AudioCommand command) {
        switch (command) {
            case BGM_STOP_ALL -> queuedBgm.values().forEach(Music::stop);
            case BGM_PAUSE_ALL -> queuedBgm.values().forEach(Music::pause);

            case BGS_STOP_ALL -> queuedBgs.values().forEach(Music::stop);
            case BGS_PAUSE_ALL -> queuedBgs.values().forEach(Music::pause);

            case SE_STOP_ALL -> queuedSe.values().forEach(Sound::stop);
            default -> throw new IllegalArgumentException("Call non-'ALL' AudioCommands with second argument.");
        }
    }

    public void dispose() {
        queuedBgm.values().forEach(Music::dispose);
        queuedBgs.values().forEach(Music::dispose);
        queuedMe.values().forEach(Sound::dispose);
        queuedSe.values().forEach(Sound::dispose);
    }

    private void playBgm(String filePath, boolean isLooping) {
        if (filePath.isEmpty()) {
            return;
        }
        Music bgm;
        if (queuedBgm.containsKey(filePath)) {
            bgm = queuedBgm.get(filePath);
        } else {
            bgm = Utils.getResourceManager().getMusicAsset(filePath);
            queuedBgm.put(filePath, bgm);
        }
        if (Utils.getSettings().isMusicOn()) {
            bgm.setLooping(isLooping);
            bgm.play();
            bgm.setVolume(BGM_VOLUME);
        } else {
            bgm.stop();
        }
    }

    private void playBgs(String filePath, boolean isLooping) {
        if (filePath.isEmpty()) {
            return;
        }
        Music bgs;
        if (queuedBgs.containsKey(filePath)) {
            bgs = queuedBgs.get(filePath);
        } else {
            bgs = Utils.getResourceManager().getMusicAsset(filePath);
            queuedBgs.put(filePath, bgs);
        }
        if (Utils.getSettings().isSoundOn()) {
            bgs.setLooping(isLooping);
            bgs.play();
            bgs.setVolume(BGS_VOLUME);
        } else {
            bgs.stop();
        }
    }

    private void playMe(String filePath, boolean isLooping) {
        Sound me;
        if (queuedMe.containsKey(filePath)) {
            me = queuedMe.get(filePath);
        } else {
            me = Utils.getResourceManager().getSoundAsset(filePath);
            queuedMe.put(filePath, me);
        }
        if (Utils.getSettings().isMusicOn()) {
            long meId = me.play();
            me.setLooping(meId, isLooping);
        } else {
            me.stop();
        }
    }

    private void playSe(AudioEvent event, boolean isLooping) {
        String filePath = event.filePath;
        float volume = event.volume;
        Sound se;
        if (queuedSe.containsKey(filePath)) {
            se = queuedSe.get(filePath);
        } else {
            se = Utils.getResourceManager().getSoundAsset(filePath);
            queuedSe.put(filePath, se);
        }
        if (Utils.getSettings().isSoundOn()) {
            long seId = se.play();
            se.setVolume(seId, volume);
            se.setLooping(seId, isLooping);
        } else {
            se.stop();
        }
    }

    private void fade(Music bgmBgs, float defaultVolume) {
        if (bgmBgs.isPlaying()) {
            float volume = bgmBgs.getVolume();
            if (volume > 0f) {
                volume -= (defaultVolume / Constant.FADE_DURATION) * Gdx.graphics.getDeltaTime();
                bgmBgs.setVolume(Math.max(volume, 0f));
            } else {
                bgmBgs.setVolume(defaultVolume);
                bgmBgs.pause();
            }
        }
    }

}

File: core/src/nl/t64/game/rpg/audio/AudioManager.kt
package nl.t64.game.rpg.audio

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.audio.Music
import com.badlogic.gdx.audio.Sound
import nl.t64.game.rpg.Utils.preferenceManager
import nl.t64.game.rpg.Utils.resourceManager
import nl.t64.game.rpg.constants.Constant
import kotlin.math.max


private const val BGM_VOLUME = 0.1f
private const val BGS_VOLUME = 0.2f

class AudioManager {

    private val queuedBgm: MutableMap<String, Music> = HashMap()
    private val queuedBgs: MutableMap<String, Music> = HashMap()
    private val queuedMe: MutableMap<String, Sound> = HashMap()
    private val queuedSe: MutableMap<String, Sound> = HashMap()

    fun toggleMusic(isMusicOn: Boolean, mustPlayBgmImmediately: Boolean) {
        if (isMusicOn && mustPlayBgmImmediately) {
            handle(AudioCommand.BGM_PLAY_LOOP, AudioEvent.BGM_TITLE)
        } else {
            handle(AudioCommand.BGM_STOP_ALL)
        }
    }

    fun toggleSound(isSoundOn: Boolean) {
        if (isSoundOn) {
            handle(AudioCommand.SE_PLAY_ONCE, AudioEvent.SE_MENU_CONFIRM)
        } else {
            handle(AudioCommand.SE_STOP, AudioEvent.SE_MENU_CONFIRM)
        }
    }

    fun possibleBgmFade(currentBgm: AudioEvent, newBgm: AudioEvent) {
        if (currentBgm !== newBgm) {
            queuedBgm.values.forEach { fade(it, BGM_VOLUME) }
        }
    }

    fun possibleBgsFade(currentBgs: List<AudioEvent>, newBgs: List<AudioEvent>) {
        currentBgs
            .filter { it != AudioEvent.NONE }
            .filter { !newBgs.contains(it) }
            .forEach { fade(queuedBgs[it.filePath]!!, BGS_VOLUME) }
    }

    fun possibleBgmSwitch(prevBgm: AudioEvent, nextBgm: AudioEvent) {
        if (prevBgm !== nextBgm) {
            handle(AudioCommand.BGM_STOP_ALL)
            handle(AudioCommand.BGM_PLAY_LOOP, nextBgm)
        }
    }

    fun possibleBgsSwitch(prevBgs: List<AudioEvent>, nextBgs: List<AudioEvent>) {
        prevBgs
            .filter { it != AudioEvent.NONE }
            .filter { !nextBgs.contains(it) }
            .forEach { handle(AudioCommand.BGS_STOP, it) }
        nextBgs
            .filter { it != AudioEvent.NONE }
            .filter { !prevBgs.contains(it) }
            .forEach { handle(AudioCommand.BGS_PLAY_LOOP, it) }
    }

    fun fadeBgmBgs() {
        queuedBgm.values.forEach { fade(it, BGM_VOLUME) }
        queuedBgs.values.forEach { fade(it, BGS_VOLUME) }
    }

    fun handle(command: AudioCommand, events: List<AudioEvent>) {
        events.forEach { handle(command, it) }
    }

    fun handle(command: AudioCommand, event: AudioEvent) {
        when (command) {
            AudioCommand.BGM_PLAY_ONCE -> playBgm(event.filePath, false)
            AudioCommand.BGM_PLAY_LOOP -> playBgm(event.filePath, true)
            AudioCommand.BGM_STOP -> queuedBgm[event.filePath]?.stop()
            AudioCommand.BGM_PAUSE -> queuedBgm[event.filePath]!!.pause()

            AudioCommand.BGS_PLAY_ONCE -> playBgs(event.filePath, false)
            AudioCommand.BGS_PLAY_LOOP -> playBgs(event.filePath, true)
            AudioCommand.BGS_STOP -> queuedBgs[event.filePath]?.stop()
            AudioCommand.BGS_PAUSE -> queuedBgs[event.filePath]!!.pause()

            AudioCommand.ME_PLAY_ONCE -> playMe(event.filePath, false)
            AudioCommand.ME_PLAY_LOOP -> playMe(event.filePath, true)
            AudioCommand.ME_STOP -> queuedMe[event.filePath]!!.stop()

            AudioCommand.SE_PLAY_ONCE -> playSe(event, false)
            AudioCommand.SE_PLAY_LOOP -> playSe(event, true)
            AudioCommand.SE_STOP -> queuedSe[event.filePath]!!.stop()
            else -> throw IllegalArgumentException("Call 'ALL' AudioCommands without second argument.")
        }
    }

    fun handle(command: AudioCommand) {
        when (command) {
            AudioCommand.BGM_STOP_ALL -> queuedBgm.values.forEach { it.stop() }
            AudioCommand.BGM_PAUSE_ALL -> queuedBgm.values.forEach { it.pause() }

            AudioCommand.BGS_STOP_ALL -> queuedBgs.values.forEach { it.stop() }
            AudioCommand.BGS_PAUSE_ALL -> queuedBgs.values.forEach { it.pause() }

            AudioCommand.SE_STOP_ALL -> queuedSe.values.forEach { it.stop() }
            else -> throw IllegalArgumentException("Call non-'ALL' AudioCommands with second argument.")
        }
    }

    fun dispose() {
        queuedBgm.values.forEach { it.dispose() }
        queuedBgs.values.forEach { it.dispose() }
        queuedMe.values.forEach { it.dispose() }
        queuedSe.values.forEach { it.dispose() }
    }

    private fun playBgm(filePath: String, isLooping: Boolean) {
        if (filePath.isEmpty()) {
            return
        }
        val bgm: Music
        if (queuedBgm.containsKey(filePath)) {
            bgm = queuedBgm[filePath]!!
        } else {
            bgm = resourceManager.getMusicAsset(filePath)
            queuedBgm[filePath] = bgm
        }
        if (preferenceManager.isMusicOn) {
            bgm.isLooping = isLooping
            bgm.play()
            bgm.volume = BGM_VOLUME
        } else {
            bgm.stop()
        }
    }

    private fun playBgs(filePath: String, isLooping: Boolean) {
        if (filePath.isEmpty()) {
            return
        }
        val bgs: Music
        if (queuedBgs.containsKey(filePath)) {
            bgs = queuedBgs[filePath]!!
        } else {
            bgs = resourceManager.getMusicAsset(filePath)
            queuedBgs[filePath] = bgs
        }
        if (preferenceManager.isSoundOn) {
            bgs.isLooping = isLooping
            bgs.play()
            bgs.volume = BGS_VOLUME
        } else {
            bgs.stop()
        }
    }

    private fun playMe(filePath: String, isLooping: Boolean) {
        val me: Sound
        if (queuedMe.containsKey(filePath)) {
            me = queuedMe[filePath]!!
        } else {
            me = resourceManager.getSoundAsset(filePath)
            queuedMe[filePath] = me
        }
        if (preferenceManager.isMusicOn) {
            val meId = me.play()
            me.setLooping(meId, isLooping)
        } else {
            me.stop()
        }
    }

    private fun playSe(event: AudioEvent, isLooping: Boolean) {
        val filePath = event.filePath
        val volume = event.volume
        val se: Sound
        if (queuedSe.containsKey(filePath)) {
            se = queuedSe[filePath]!!
        } else {
            se = resourceManager.getSoundAsset(filePath)
            queuedSe[filePath] = se
        }
        if (preferenceManager.isSoundOn) {
            val seId = se.play()
            se.setVolume(seId, volume)
            se.setLooping(seId, isLooping)
        } else {
            se.stop()
        }
    }

    private fun fade(bgmBgs: Music, defaultVolume: Float) {
        if (bgmBgs.isPlaying) {
            var volume = bgmBgs.volume
            if (volume > 0f) {
                volume -= defaultVolume / Constant.FADE_DURATION * Gdx.graphics.deltaTime
                bgmBgs.volume = max(volume, 0f)
            } else {
                bgmBgs.volume = defaultVolume
                bgmBgs.pause()
            }
        }
    }

}
